### DISCUSSION TITLE
managing / versioning secrets - how do you do it?

### DISCUSSION THREAD

**Discussion by mamema** [authelia/authelia#6729] [General]
Hi,

i'm not asking about the technical approach authelia delivers, which works great with the _FILE=/secretpath approach

My question leans more into the fact that the secerts (i'm using docker swarm) are entered/created and can only be deleted.

How to you manage the pletora of secrets in an environment, not only authelia specific. Do you save them additionaly elsewhere? Offsite? In clear text in an file ? :-)

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

Managing secrets in any environment, especially in complex systems like Docker Swarm, Kubernetes, or any other orchestration platform, is a critical aspect of maintaining security and operational integrity. While Authelia provides a robust method for handling secrets within its scope, the broader question of managing a plethora of secrets across various systems and services is a common challenge.

Here are some general strategies for managing secrets across environments, not limited to Authelia:

1. **Secrets Management Tools**: Utilize dedicated secrets management tools like HashiCorp Vault, AWS Secrets Manager, Azure Key Vault, or Google Cloud Secret Manager. These tools are designed to securely store, access, and manage secrets across environments. They often provide features like secret rotation, access control policies, and audit logs.

2. **Environment-Specific Secrets Storage**: For container orchestration systems like Docker Swarm or Kubernetes, leverage their native secrets management capabilities. For example, Kubernetes Secrets or Docker Swarm Secrets provide a way to securely store and manage sensitive data within those environments.

3. **Encryption**: If you must store secrets outside of these managed solutions (e.g., for backup purposes), ensure they are encrypted at rest. Tools like GPG can be used to encrypt files containing secrets before storing them in a secure location.

4. **Access Control**: Limit access to secrets as much as possible. Use role-based access control (RBAC) to ensure only the necessary personnel or systems have access to specific secrets. This minimizes the risk of unauthorized access.

5. **Audit and Rotate Secrets Regularly**: Regularly audit your secrets for exposure risks and rotate them periodically to minimize the impact of potential leaks. Automated tools and secrets management solutions often provide mechanisms to facilitate secret rotation.

6. **Secure Backup Practices**: For offsite or backup storage of secrets, ensure that the storage medium is secure and encrypted. Access to these backups should be tightly controlled and monitored.

7. **Documentation and Policies**: Maintain clear documentation and policies regarding secrets management. This includes procedures for creating, storing, rotating, and revoking secrets. Ensure that team members are trained on these policies.

8. **Avoid Storing Secrets in Source Code**: Never store secrets in source code or version control systems, even in private repositories. Always use environment variables or secrets management tools to inject secrets into your applications.

9. **Regularly Review and Update Security Practices**: The landscape of security threats is constantly evolving. Regularly review and update your secrets management practices to address new threats and incorporate best practices.

While Authelia's documentation and community discussions provide insights into managing secrets within Authelia, these broader strategies can help manage secrets across your entire environment. Remember, the goal is to minimize the risk of unauthorized access to sensitive information while maintaining operational efficiency.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Reply by mamema** [reply]
yes AI, you tell me i should use docker swarm. Hmm.... instead of A ritifical I ntelligence i would like to have A dditionally I ntelligence from humans

**Comment by mamema** [comment]
i've googled a bit and learned, with a kubernetes cluster there is decoding of secrets possible...
Nothing in docker swarm, everyone should connect the docker container and make a cat /run/secrets/*
This is it? Really?
This isn't production ready.
So if i want / have to change a password secret i scan <100 compose files to find out ah.... container_xyz is using my secret. 
Or i try to delete the secret and fast read the error which container is using my secret.

Are YOU working like this?

**Reply by ivanov17** [reply]
If you want to manage secrets, you can manage them using Ansible. For example, store them as variables using Ansible Vault, or just generate them once. Authelia uses secrets, but doesn't manage them.

**Reply by mamema** [reply]
because of that i wrote in the general area. Secret managment is not an authelia topic, it's an broader issue. Are you using ansible vault?

**Reply by ivanov17** [reply]
Yes, I use Ansible roles to manage Podman containers, including the Authelia container. It allows me to create persistent volumes before creating containers, use templates for configuration files, and store some secret values in encrypted files using Ansible Vault. Also, I can store the name of the secret as a variable and then mount that secret to any other container if needed.
